node/actors/executor/src/attestation.rs

```diff
@@ -39,51 +38,68 @@ impl AttesterRunner {
+        let mut last_batch_number = self
+            .batch_store
+            .last_batch_number(ctx)
+            .await
+            .context("last_batch_number")?
+            .unwrap_or_default();
 
-            for bn in unsigned_batch_numbers {
-                // If we have already voted on this we can move on, no need to fetch the payload again.
-                // Batches appear in the store in order, even if we might have QC for a newer and not for an older batch,
-                // so once published our vote for a certain height, we can expect that we only have to vote on newer batches.
-                if bn < min_batch_number {
-                    continue;
-                }
+        let mut earliest_batch_number = self
+            .batch_store
+            .earliest_batch_number_to_sign(ctx)
+            .await
+            .context("earliest_batch_number_to_sign")?
+            .unwrap_or(last_batch_number);
 
-                if let Some(batch) = self
+        loop {
+            while earliest_batch_number <= last_batch_number {
+                // Try to get the next batch to sign; the commitment might not be available just yet.
+                let Some(batch) = self
                     .batch_store
-                    .batch_to_sign(ctx, bn)
+                    .batch_to_sign(ctx, earliest_batch_number)
                     .await
                     .context("batch_to_sign")?
-                {
-                    min_batch_number = batch.number.next();
+                else {
+                    break;
+                };
+
+                // The certificates might be collected out of order because of how gossip works;
+                // we could query the DB to see if we already have a QC, or we can just go ahead
+                // and publish our vote, and let others ignore it.
 
-                    self.publisher
-                        .publish(attesters, &self.attester.key, batch)
-                        .await
-                        .context("publish")?;
-                }
+                // We only have to publish a vote once; future peers can pull it from the register.
+                self.publisher
+                    .publish(attesters, &self.attester.key, batch)
+                    .await
+                    .context("publish")?;
+
+                earliest_batch_number = earliest_batch_number.next();
             }
 
+            // Wait some time before we poll the database again to see if there is a new batch to sign.
             ctx.sleep(POLL_INTERVAL).await?;
+
+            // Refresh the upper end of the range.
+            if earliest_batch_number > last_batch_number {
+                last_batch_number = self
+                    .batch_store
+                    .last_batch_number(ctx)
+                    .await
+                    .context("last_batch_number")?
+                    .unwrap_or(last_batch_number);
+            }
```

Comment:
it's indeed easier to follow now, but I find the execution flow here peculiar... 

We maintain these two pointers in memory (`mut earliest_batch_number` and `mut last_batch_number`), we go through batches in between those two numbers in a `while` loop and sign them, until either:
1. needed batch commitment is missing in the DB (we `break`)
2. we reached `last_batch_number + 1` (`while` condition doesn't hold true)

then we sleep for `POLL_INTERVAL` and conditionally bump `last_batch_number` - only if condition `2.` from above triggered - otherwise we return to the loop, waiting for `batch_to_sign` to become `batch_to_sign(ctx, earliest_batch_number)`. Why do we need both? We essentially have two restrictions that prevent event loop from continuing:
* `batch_to_sign(ctx, earliest_batch_number).is_some()`
* `last_batch_number(ctx)`

we check them in a different way and in different places, and also treat them differently - when (1) is false, we poll, whereas when we do (2), we update `last_batch_number`, but they do effectively the same - makes this component wait.

Could we achieve the same result in a way that's easier to follow? Something like:
```
let last_persisted_batch = self
                    .batch_store
                    .last_batch_number(ctx)
                    .await
                    .context("last_batch_number")?
                    .unwrap_or(last_batch_number);
let mut current_batch_number = self
            .batch_store
            .earliest_batch_number_to_sign(ctx)
            .await
            .context("earliest_batch_number_to_sign")?
            .unwrap_or(last_persisted_batch);
          
// log the current and last_persisted_batch numbers

loop {
             let batch = get_batch_to_sign_blocking(current_batch_number);
             // log the batch number and time spent waiting (although will be apparent from log lines)
            // also consider sending this timing to prometheus
             
            self.publisher
                    .publish(attesters, &self.attester.key, batch)
                    .await
                    .context("publish")?;
                  
           current_batch_number = current_batch_number.next()
}

// blocks until the batch is available in the DB for signing
    fn get_batch_to_sign_blocking(batch_number) -> attester::Batch {
    loop {
        let res =self
                    .batch_store
                    .batch_to_sign(ctx, batch_number)
                    .await
                    .context("batch_to_sign")?
        match res {
            Some(batch) => {
                return batch;
            }
            None => {
                sleep(polling_interval);
            }
        }
    }
}
```
the blocking logic can be done in `PersistentBatchStore` directly - will need to look deeper to see what makes more sense but I think we have a lot of polling there already.

Am I missing something or is the code above equivalent? If so, I like it more since the control flow is very clear (albeit still imperative). State space is much smaller - the only mutable var is `current_batch_number`.

Reply:
Yours looks equivalent in what it achieves, just doing different queries in the background. I'll rewrite it that way.

I'd rather avoid blocking inside the `PersitentBatchStore` by polling inside the call. Easier to compose if individual methods do just simple things.

Reply:
Done in https://github.com/matter-labs/era-consensus/pull/150/commits/49a9b94d2d8b76e7713bc6be3edf7fc3ad97ac84

I didn't add timing info - it's going to be on average the same as any recurring event between batches - but I'll add metrics as part of BFT-486 if that's okay.

Reply:
> Yours looks equivalent in what it achieves, just doing different queries in the background

for me the important part here is that we never invoke `last_batch_number` (less things to keep in mind) and instead of two mutable variables just stores one. The state in the original version is represented by a `range`, whereas here it's just one number. For me it's substantially easier to reason about, do you agree?

Reply:
Yes I agree it's easier to follow, and thank you dedicating you time to make it nicer! 

I did it in a range because that's what you asked [earlier](https://github.com/matter-labs/era-consensus/pull/148#discussion_r1668747408): 

> Could we maybe keep a mutable range instead and operate over it? And thus only request the last_persisted_batch from batch_store to update the upper limit.

I just focussed on making that happen without looking for further opportunities 😅

Reply:
oh yes agree could have proposed this the first time 👍